README.md

```diff
@@ -1,189 +1,180 @@
+- Implementations can work fully on a per-file basis, with no need for cross-file knowledge.
+- No new namespace or type of second-class value is added--decorators are functions.
 
-### `@logged`
+# Examples
 
-The `@logged` decorator logs a console message when a method starts and finishes. Many other popular decorators will also want to wrap a function, e.g., `@deprecated`, `@debounce`, `@memoize`, dependency injection, etc.
+A few examples of how to implement and use decorators in this proposal:
+
+## `@logged`
+
+The `@logged` decorator logs a console message when a method starts and finishes. Many other popular decorators will also want to wrap a function, e.g., `@deprecated`, `@debounce`, `@memoize`, etc.
 
 Usage:
 
 ```mjs
-import { @logged } from "./logged.mjs";
+import { logged } from "./logged.mjs";
 
 class C {
   @logged
-  method(arg) {
+  m(arg) {
     this.#x = arg;
   }
 
   @logged
   set #x(value) { }
 }
 
-new C().method(1);
-// starting method with arguments 1
+new C().m(1);
+// starting m with arguments 1
 // starting set #x with arguments 1
 // ending set #x
-// ending method
+// ending m
 ```
 
-`@logged` can be implemented in JavaScript in terms of built-in decorators: The `@logged` decorator is defined in terms of the `@wrap` built-in decorator, which takes a function as an argument. The method is passed through this function to get the method which is finally present on the class. `@wrap` is similar to core-decorators' [`@decorate`](https://www.npmjs.com/package/core-decorators#decorate) decorator.
+`@logged` can be implemented in JavaScript as a decorator. Decorators are functions that are called with an argument containing what's being decorated. For example:
+- A decorated method is called with the method being decorated
+- A decorated getter is called with the getter function being decorated
+- A decorated setter is called with the setter function being decorated
+
+(Decorators are called with a second parameter giving more context, but we don't need those details for the `@logged` decorator.)
+
+The return value of a decorator is a new value that replaces the thing it's wrapping. For methods, getters and setters, the return value is another function to replace that method, getter or setter.
 
 ```mjs
 // logged.mjs
 
-export decorator @logged {
-  @wrap(f => {
-    const name = f.name;
-    function wrapped(...args) {
-      console.log(`starting ${name} with arguments ${args.join(", ")}`);
-      f.call(this, ...args);
-      console.log(`ending ${name}`);
-    }
-    wrapped.name = name;
-    return wrapped;
-  })
+export function logged(f) {
+  const name = f.name;
+  function wrapped(...args) {
+    console.log(`starting ${name} with arguments ${args.join(", ")}`);
+    const ret = f.call(this, ...args);
+    console.log(`ending ${name}`);
+    return ret;
+  }
+  Object.defineProperty(wrapped, 'name', { value: name, configurable: true })
+  return wrapped;
```

Comment:
In my humbel opinion the function's name explicit definition is unnnecesary and litle bit confuse. Can you create a new method with other name if change the function name. I undestand that not.

Reply:
Yes, changing the name is just an optional thing if you want to make reflection/debugging/metaprogramming nicer.

Reply:
Yes, changing the name is just an optional thing if you want to make reflection/debugging/metaprogramming nicer.